src/v/archival/segment_reupload.cc

```diff
@@ -87,7 +114,25 @@ void segment_collector::do_collect(segment_collector_mode mode) {
         }
 
         auto segment_size = result.segment->size_bytes();
-        if (_collected_size + segment_size > _max_uploaded_segment_size) {
+        if (
+          _target_end_inclusive.has_value()
+          && result.segment->offsets().committed_offset
+               >= _target_end_inclusive.value()) {
```

Comment:
nit: it kind of feels like this condition should be baked into the while condition by comparing `current_segment_end` and _target_end_inclusive. At least, it'd get avoid having to reason about what this branch is doing mid-loop, and maybe makes it easier to follow (not sure, but here's a stab):

```
auto target_end_inclusive = _target_end_inclusive.value_or(_manifest.get_last_offset());
bool strict_size_bound = !_target_end_inclusive.has_value();
while (current_segment_end < target_end_inclusive) {
   ...
    if (strict_size_bound && _collected_size + segment_size > _max_uploaded_segment_size) {
        ...
        break
    }
    ...
    current_segment_end = result.segment->offsets().committed_offset;
    ....
}
```